Azure/Azure-AppServices-Diagnostics-Portal
# Curibe/opt insights integration diag solve

## Overview

Adding integration with Code Optimizations (previously known as Optimization Insights), to display under Diagnose & Solve problems

Code Optimizations team wants us to show if customers have insights for their apps when they go to D&SP blade for their Web App.  The majority of their customers will run on Web Apps and it will be important for customers to know when troubleshooting their Web App if there are Code Optimizations insights.
## Related Work Item

Feature 3007
https://dev.azure.com/app-service-diagnostics-portal/app-service-diagnostics-portal/_workitems/edit/3007
## Type of change




## Solution description

### This PR:
opt-insights-enablement.component.ts
It uses the Optinsights.service.ts to query Code Optimizations to see if there are insights for this resource.
Optinsights.service.ts 
It requests a scoped portal token used to query Code Optimizations token api, used to retrieve an OAuth token that in turn will allow us to query Code Optimizations service to retrieve the insights.

## How Can This Be Tested? <span>&#128269;</span>

1. Go to a Web App integrated with Application Insights which has profilers enabled and Code Optimization Insights.
2. Click on Diagnose and Solve problems then in Availability and Performance. This will load the Overview detector.
3.  In the Overview detector first it will try to show Application Insights. If not connected, it will prompt to connect to integrate with Application Insights. Once that's done it will load below the Code Optimizations section:
![image](https://user-images.githubusercontent.com/7376302/234368276-5e9579d3-1237-40e6-9fd9-58092811d8d8.png)

Web App with results:
https://ms.portal.azure.com/?websitesextension_ext=asd.env%3Dstaging#view/WebsitesExtension/SCIFrameBlade/id/%2Fsubscriptions%2Fbbe41737-1ade-44df-8e33-217f11b8b452%2FresourceGroups%2Fdiagservice%2Fproviders%2FMicrosoft.Web%2Fsites%2Fdiagservicecw/categoryId/AvailabilityAndPerformanceWindows/optionalParameters~/%5B%7B%22key%22%3A%22categoryId%22%2C%22value%22%3A%22AvailabilityAndPerformanceWindows%22%7D%5D
Web App connecting to Code Optimizations with no results:
https://ms.portal.azure.com/?websitesextension_ext=asd.env%3Dstaging#view/WebsitesExtension/SCIFrameBlade/id/%2Fsubscriptions%2Fa2c49f0f-548f-4f3b-86a4-39c8eafc0780%2FresourceGroups%2Foptinsights%2Fproviders%2FMicrosoft.Web%2Fsites%2Fbishoptrouble/categoryId/AvailabilityAndPerformanceWindows/optionalParameters~/%5B%7B%22key%22%3A%22categoryId%22%2C%22value%22%3A%22AvailabilityAndPerformanceWindows%22%7D%5D
Web App showing error connecting to Code Optimizations:
https://ms.portal.azure.com/?websitesextension_ext=asd.env%3Dstaging#view/WebsitesExtension/SCIFrameBlade/id/%2Fsubscriptions%2Fdd5a12e0-3f90-47b8-9c07-7f9f0c1af13a%2FresourceGroups%2Fdefault-web-centralus%2Fproviders%2FMicrosoft.Web%2Fsites%2Fbathwater/categoryId/AvailabilityAndPerformanceWindows/optionalParameters~/%5B%7B%22key%22%3A%22categoryId%22%2C%22value%22%3A%22AvailabilityAndPerformanceWindows%22%7D%5D
Web App not applicable (it should show nothing):
https://ms.portal.azure.com/?websitesextension_ext=asd.env%3Dstaging#view/WebsitesExtension/SCIFrameBlade/id/%2Fsubscriptions%2Fa2c49f0f-548f-4f3b-86a4-39c8eafc0780%2FresourceGroups%2Frr-lnx-rg%2Fproviders%2FMicrosoft.Web%2Fsites%2Frr-lnx-1/categoryId/AvailabilityAndPerformanceLinux/optionalParameters~/%5B%7B%22key%22%3A%22categoryId%22%2C%22value%22%3A%22AvailabilityAndPerformanceLinux%22%7D%5D
## Checklist <span>&#9989;</span>
- [ ] I have ensured that my changes support accessibility and can be accessed with the keyboard alone //pending for next deployment

## Extra Notes
